lib/devise/strategies/passwordless_authenticatable.rb

```diff
@@ -0,0 +1,29 @@
+require "devise/strategies/authenticatable"
+require_relative "../../../app/mailers/user_mailer"
+
+module Devise
+  module Strategies
+    class PasswordlessAuthenticatable < Authenticatable
+      def authenticate!
+        if params[:user].present?
+          user = User.find_by(email: params[:user][:email])
+
+          if user&.update(
+            login_token: SecureRandom.hex(10),
+            login_token_valid_until: 60.minutes.from_now)
+
+            url = Rails.application.routes.url_helpers.email_confirmation_url(
+              login_token: user.login_token,
+              host: Rails.application.config.domain,
+            )
+
+            UserMailer.sign_in_email(user, url).deliver_now
+            fail!("An email was sent to you with a magic link.")
+          end
```

Comment:
I find it a bit hard to understand what's going on here. 

If the user exists, we try to update their login token. And if that succeeds we send them an email, and we `fail!` with a confirmation message? That seems like a hack. If it's a standard way to do it, I would add a comment with link to devise documentation. 

And what happens when the user does not exist? Devise magic catches that afterwards? Beforehand? 

Can we do something like https://gist.github.com/r00k/906356 does? With `success!` method for the good case, and a `fail` for the bad case?

Reply:
>> If the user exists, we try to update their login token. And if that succeeds we send them an email, and we fail! with a confirmation message? That seems like a hack.

The fail! is because we haven't logged in the user, we have just sent them the magic link. success! would try and set the cookie etc which is not what we need.

Reply:
Alright, that makes sense.

We will probably revisit this sooner or later (when designers get their hands on the login flow I imagine), but it can probably stay this way for time being.